File: app/src/main/java/org/mesonet/app/radar/RadarFragment.java
package org.mesonet.app.radar;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.content.Context;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.design.widget.BaseTransientBottomBar;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentTransaction;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroup;
import android.widget.SeekBar;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;

import org.mesonet.app.R;
import org.mesonet.app.baseclasses.BaseFragment;
import org.mesonet.app.databinding.RadarFragmentBinding;
import org.mesonet.app.filterlist.FilterListFragment;

import java.util.Observable;
import java.util.Observer;

import javax.inject.Inject;


public class RadarFragment extends BaseFragment implements GoogleMapController.GoogleMapProvider, FilterListFragment.FilterListCloser
{
    @Inject
    GoogleMapController mMapController;


    RadarFragmentBinding mBinding;

    private Snackbar mSnackbar;



    private BindingReadyListener mBindingReadyListener;



    @Override
    public void onCreate(Bundle inSavedInstanceState)
    {
        super.onCreate(inSavedInstanceState);

        mMapController.SetProvider(this);
    }



    @Override
    public void onActivityCreated(Bundle inSavedInstanceState)
    {
        super.onActivityCreated(inSavedInstanceState);

        mBinding.map.onCreate(inSavedInstanceState);
    }



    @Override
    public View onCreateView(LayoutInflater inInflater, ViewGroup inContainer, Bundle inSavedInstanceState)
    {
        mBinding = DataBindingUtil.inflate(inInflater, R.layout.radar_fragment, inContainer, false);

        if(mBindingReadyListener != null)
            mBindingReadyListener.BindingReady(mBinding);

        mBindingReadyListener = null;

        mSnackbar = Snackbar.make(mBinding.radarLayout, "", Snackbar.LENGTH_INDEFINITE).setAction("Change", new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentTransaction transaction = getChildFragmentManager().beginTransaction();
                transaction.replace(R.id.childFragmentContainer, new FilterListFragment());
                transaction.commit();
                RevealView(mBinding.radarLayout);
                mBinding.playPauseButton.hide();
            }
        });

        mBinding.playPauseButton.SetPlayPauseState(mMapController.GetPlayPauseState());
        mBinding.playPauseButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mMapController.TogglePlay();
            }
        });

        mBinding.transparencySeekBar.setMax(255);
        mBinding.transparencySeekBar.setProgress(Math.round(255.0f * (1 - mMapController.GetTransparency())));
        mBinding.transparencySeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar inSeekBar, int inProgress, boolean inFromUser) {
                mMapController.SetTransparency((1 - (inProgress / 255.0f)));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        mSnackbar.show();

        return mBinding.getRoot();
    }



    @Override
    public void onResume()
    {
        super.onResume();

        mBinding.map.onResume();
    }



    @Override
    public void onPause()
    {
        mBinding.map.onPause();

        super.onPause();
    }



    @Override
    public void onDestroy()
    {
        mMapController.SetProvider(null);
        mBinding.map.onDestroy();

        super.onDestroy();
    }



    @Override
    public void onSaveInstanceState(Bundle inSaveInstanceState)
    {
        super.onSaveInstanceState(inSaveInstanceState);

        mBinding.map.onSaveInstanceState(inSaveInstanceState);
    }



    @Override
    public void onLowMemory()
    {
        super.onLowMemory();

        mBinding.map.onLowMemory();
    }



    @Override
    public void GetMap(final GoogleMapListener inListener)
    {
        BindingReadyListener bindingReadyListener = new BindingReadyListener() {
            @Override
            public void BindingReady(RadarFragmentBinding inBinding) {
                if(inListener != null && inBinding != null) {
                    inBinding.map.getMapAsync(new OnMapReadyCallback() {
                        @Override
                        public void onMapReady(GoogleMap inMap) {
                            inListener.MapFound(getContext(), inMap);
                        }
                    });
                }
            }
        };

        if(mBinding == null)
            mBindingReadyListener = bindingReadyListener;

        else
            bindingReadyListener.BindingReady(mBinding);
    }



    @Override
    public void SetTime(String inTimeString)
    {
        if(mSnackbar != null) {
            String timeString = "";

            if(inTimeString != null && !inTimeString.isEmpty()) {
                timeString = "\n" + inTimeString;
                mBinding.playPauseButton.show();
            }
            else
                mBinding.playPauseButton.hide();

            mSnackbar.setText(mMapController.GetRadarName() + timeString);
        }
    }

    @Override
    public void Close() {
        RevealView(mBinding.radarLayout);
        mSnackbar.show();
        mBinding.playPauseButton.show();
    }



    private void RevealView(View view)
    {
        final int cx = (view.getLeft() + view.getRight()) - 24;
        final int cy = (view.getTop() + view.getBottom()) / 2;
        final float radius = Math.max(mBinding.childFragmentContainer.getWidth(), mBinding.childFragmentContainer.getHeight()) * 2.0f;

        if (mBinding.childFragmentContainer.getVisibility() != View.VISIBLE) {
            mBinding.childFragmentContainer.setVisibility(View.VISIBLE);
            final Animator hide = ViewAnimationUtils.createCircularReveal(mBinding.childFragmentContainer, cx, cy, 0, radius);
            hide.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    mBinding.childFragmentContainer.setVisibility(View.VISIBLE);
                    hide.removeListener(this);
                }
            });
            hide.start();

        } else {
            final Animator reveal = ViewAnimationUtils.createCircularReveal(
                    mBinding.childFragmentContainer, cx, cy, radius, 0);
            reveal.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    mBinding.childFragmentContainer.setVisibility(View.INVISIBLE);
                    reveal.removeListener(this);
                }
            });
            reveal.start();
        }
    }


    private interface BindingReadyListener
    {
        void BindingReady(RadarFragmentBinding inBinding);
    }
}

File: app/src/main/java/org/mesonet/app/radar/RadarFragment.kt
package org.mesonet.app.radar

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.app.Activity
import android.content.res.Configuration
import android.databinding.DataBindingUtil
import android.graphics.Bitmap
import android.os.Build
import android.os.Bundle
import android.text.Html
import android.view.*
import android.widget.SeekBar
import com.mapbox.mapboxsdk.camera.CameraPosition
import com.mapbox.mapboxsdk.geometry.LatLng
import com.mapbox.mapboxsdk.geometry.LatLngQuad
import com.mapbox.mapboxsdk.maps.MapboxMapOptions
import com.mapbox.mapboxsdk.maps.Style
import com.mapbox.mapboxsdk.maps.SupportMapFragment
import com.mapbox.mapboxsdk.style.layers.PropertyFactory
import com.mapbox.mapboxsdk.style.layers.RasterLayer
import com.mapbox.mapboxsdk.style.sources.ImageSource
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.Observables
import org.mesonet.app.R

import org.mesonet.app.baseclasses.BaseFragment
import org.mesonet.app.databinding.RadarMapFragmentBinding
import org.mesonet.app.filterlist.FilterListFragment
import org.mesonet.dataprocessing.PageStateInfo
import org.mesonet.dataprocessing.radar.MapboxMapController
import org.mesonet.dataprocessing.radar.RadarDataController
import org.mesonet.dataprocessing.radar.RadarImageDataProvider
import org.mesonet.dataprocessing.userdata.Preferences
import org.mesonet.models.radar.RadarDetails

import javax.inject.Inject
import kotlin.math.roundToInt


class RadarFragment : BaseFragment(), FilterListFragment.FilterListCloser {

    private val kRasterImageName = "Radar"
    private val kLightThemeUrl = "mapbox://styles/okmesonet/cjkmvlllh7oo12rplp8zufopc"
    private val kDarkThemeUrl = "mapbox://styles/okmesonet/cjkjtvw1n0nnm2rpen1ggwh99"

    @Inject
    internal lateinit var mRadarImageDataProvider: RadarImageDataProvider

    @Inject
    internal lateinit var mPreferences: Preferences

    @Inject
    internal lateinit var mMapController: MapboxMapController

    @Inject
    internal lateinit var mActivity: Activity

    private var mTransparencyDisposable: Disposable? = null
    private var mSelectedRadarDisposable: Disposable? = null
    private var mActiveImageDisposable: Disposable? = null
    private var mImageListDisposable: Disposable? = null
    private var mCameraUpdateDisposable: Disposable? = null
    private var mPageStateDisposable: Disposable? = null

    private var mEmtpyImage = Bitmap.createBitmap(1, 1, Bitmap.Config.ALPHA_8)

    private var mBinding: RadarMapFragmentBinding? = null

    private var mMapFragment: SupportMapFragment? = null

    private var mRadarLayer: RasterLayer? = null
    private var mRadarImageSource: ImageSource? = null

    private var mLastRadarUpdated = ""
    private var mLastColorTheme = Preferences.RadarColorThemePreference.kLight

    private var mHoldPlay = false
    private var mState = PageStateInfo.PageState.kLoading


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mMapController.GetCameraPositionObservable().observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<CameraPosition>
        {
            override fun onComplete() {}
            override fun onSubscribe(d: Disposable)
            {
                mCameraUpdateDisposable = d
            }

            override fun onNext(t: CameraPosition) {
                mMapFragment?.getMapAsync { map ->
                    map.cameraPosition = t
                }
            }

            override fun onError(e: Throwable) {
                e.printStackTrace()
            }
        })

    }


    override fun onCreateView(inInflater: LayoutInflater, inContainer: ViewGroup?, inSavedInstanceState: Bundle?): View {
        mBinding = DataBindingUtil.inflate(inInflater, R.layout.radar_map_fragment, inContainer, false)

        context?.let { fragmentContext ->
            val options = MapboxMapOptions.createFromAttributes(fragmentContext)
            options.textureMode(true)
            options.logoGravity(Gravity.TOP or Gravity.START)
            options.attributionGravity(Gravity.TOP or Gravity.START)

            mMapFragment = SupportMapFragment.newInstance(options)

            mRadarImageSource = ImageSource(kRasterImageName, LatLngQuad(LatLng(0.0, 0.0),
                    LatLng(0.0, 0.0),
                    LatLng(0.0, 0.0),
                    LatLng(0.0, 0.0)), R.drawable.blank_drawable)


            mRadarLayer = RasterLayer(kRasterImageName, mRadarImageSource?.id)

            mMapFragment?.let {
                val mapTransaction = childFragmentManager.beginTransaction()
                mapTransaction.add(R.id.mapContainer, it)
                mapTransaction.commit()
            }

            mBinding?.playPauseButton?.SetPlayPauseState(mMapController.GetPlayPauseStateObservable())

            mBinding?.playPauseButton?.setOnClickListener {
                mMapController.TogglePlay()
            }

            mBinding?.readingInfoLayout?.setOnClickListener {
                val filterTransaction = childFragmentManager.beginTransaction()
                filterTransaction.replace(R.id.childFragmentContainer, FilterListFragment())
                filterTransaction.commit()
                RevealView(mBinding?.readingInfoLayout, mBinding?.childFragmentContainer)
                (mBinding?.playPauseButton as View?)?.visibility = View.GONE
            }

            mBinding?.opacityDrawerButtonLayout?.setOnClickListener {
                RevealView(mBinding?.opacityDrawerButtonLayout, mBinding?.opacityLayout)
            }

            mBinding?.opacityCloseButton?.setOnClickListener {
                RevealView(mBinding?.opacityDrawerButtonLayout, mBinding?.opacityLayout)
            }

            mBinding?.transparencySeekBar?.max = 255

            mMapController.GetTransparencySubject().observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<Float> {
                var disposable: Disposable? = null
                override fun onComplete() {}

                override fun onSubscribe(d: Disposable) {
                    disposable = d
                }

                override fun onNext(t: Float) {
                    mBinding?.transparencySeekBar?.progress = (t * 255.0f).roundToInt()
                    disposable?.dispose()
                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                }

            })

            mBinding?.transparencySeekBar?.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                override fun onProgressChanged(inSeekBar: SeekBar, inProgress: Int, inFromUser: Boolean) {
                    val result = inProgress / 255.0f

                    val subject = mMapController.GetTransparencySubject()

                    if (!subject.hasValue() || subject.value != result)
                        mMapController.GetTransparencySubject().onNext(result)
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {

                }

                override fun onStopTrackingTouch(seekBar: SeekBar) {

                }
            })
        }

        return mBinding?.root?: View(context)
    }


    override fun onResume() {
        super.onResume()

        val myContext = context

        if(myContext != null)
            mRadarImageDataProvider.OnResume(myContext)

        if(mPageStateDisposable?.isDisposed != false) {
            mRadarImageDataProvider.GetPageStateObservable().observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<PageStateInfo> {
                override fun onComplete() {}
                override fun onSubscribe(d: Disposable) {
                    mPageStateDisposable = d
                }

                override fun onNext(t: PageStateInfo) {
                    mState = t.GetPageState()
                    when (t.GetPageState()) {
                        PageStateInfo.PageState.kLoading -> {
                            mTransparencyDisposable?.dispose()
                            mSelectedRadarDisposable?.dispose()
                            mRadarLayer?.setProperties(PropertyFactory.rasterOpacity(0.0f))
                            (mBinding?.playPauseButton as View?)?.visibility = View.GONE
                            mBinding?.readingInfoLayout?.visibility = View.GONE
                        }
                        PageStateInfo.PageState.kError -> {
                            mTransparencyDisposable?.dispose()
                            mSelectedRadarDisposable?.dispose()
                            mRadarLayer?.setProperties(PropertyFactory.rasterOpacity(0.0f))
                            (mBinding?.playPauseButton as View?)?.visibility = View.GONE
                            val errorMessage = t.GetErrorMessage()

                            if (errorMessage != null) {
                                mBinding?.readingInfoTextView?.text = errorMessage
                                mBinding?.readingInfoLayout?.visibility = View.VISIBLE
                            } else
                                mBinding?.readingInfoLayout?.visibility = View.GONE
                        }
                        PageStateInfo.PageState.kData -> {
                            mMapController.GetTransparencySubject().observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<Float> {
                                override fun onComplete() {}
                                override fun onSubscribe(d: Disposable) {
                                    mTransparencyDisposable?.dispose()
                                    mTransparencyDisposable = d
                                }

                                override fun onNext(t: Float) {
                                    mRadarLayer?.setProperties(PropertyFactory.rasterOpacity(t))
                                }

                                override fun onError(e: Throwable) {
                                    onNext(0.0f)
                                    e.printStackTrace()
                                }
                            })

                            Observables.combineLatest(mRadarImageDataProvider.GetSiteInfoObservable(),
                                                    mPreferences.RadarColorThemePreferenceObservable(mActivity)).observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<Pair<Map.Entry<String, RadarDetails>, Preferences.RadarColorThemePreference>> {
                                override fun onComplete() {}

                                override fun onSubscribe(d: Disposable) {
                                    mSelectedRadarDisposable?.dispose()
                                    mSelectedRadarDisposable = d
                                }

                                override fun onError(e: Throwable) {
                                    e.printStackTrace()

                                    SetSnackbarText("Error", "", "")
                                }

                                override fun onNext(radarInfo: Pair<Map.Entry<String, RadarDetails>, Preferences.RadarColorThemePreference>) {
                                    if (mLastRadarUpdated != radarInfo.first.key || mLastColorTheme != radarInfo.second) {

                                        mLastRadarUpdated = radarInfo.first.key
                                        mLastColorTheme = radarInfo.second

                                        mHoldPlay = true

                                        if(mState == PageStateInfo.PageState.kData) {
                                            SetSnackbarText(radarInfo.first.key, radarInfo.first.value.GetName(), "")
                                            mBinding?.readingInfoLayout?.visibility = View.VISIBLE
                                        }

                                        mMapFragment?.getMapAsync { map ->

                                            if(map.style == null) {
                                                map.uiSettings.attributionDialogManager
                                                map.uiSettings.isRotateGesturesEnabled = false
                                                map.uiSettings.isTiltGesturesEnabled = false

                                                var styleUrl = kLightThemeUrl

                                                if (radarInfo.second == Preferences.RadarColorThemePreference.kDark)
                                                    styleUrl = kDarkThemeUrl

                                                map.setStyle(Style.Builder().fromUri(styleUrl))
                                            }

                                            map.getStyle() { style ->
                                                val sourceId = mRadarImageSource?.id ?: ""
                                                val layerId = mRadarLayer?.id ?: ""

                                                val radarImageSource = mRadarImageSource
                                                val radarLayer = mRadarLayer

                                                if (sourceId.isNotEmpty() && style.getSource(sourceId) == null && radarImageSource != null)
                                                    style.addSource(radarImageSource)
                                                if (radarInfo.first.key.isNotEmpty() && style.getLayer(layerId) == null && radarLayer != null)
                                                    style.addLayer(radarLayer)
                                                mRadarImageSource?.setCoordinates(LatLngQuad(LatLng(radarInfo.first.value.GetNorthEastCorner()?.GetLatitude()?.toDouble()
                                                        ?: 0.0, radarInfo.first.value.GetSouthWestCorner()?.GetLongitude()?.toDouble()
                                                        ?: 0.0),
                                                        LatLng(radarInfo.first.value.GetNorthEastCorner()?.GetLatitude()?.toDouble()
                                                                ?: 0.0, radarInfo.first.value.GetNorthEastCorner()?.GetLongitude()?.toDouble()
                                                                ?: 0.0),
                                                        LatLng(radarInfo.first.value.GetSouthWestCorner()?.GetLatitude()?.toDouble()
                                                                ?: 0.0, radarInfo.first.value.GetNorthEastCorner()?.GetLongitude()?.toDouble()
                                                                ?: 0.0),
                                                        LatLng(radarInfo.first.value.GetSouthWestCorner()?.GetLatitude()?.toDouble()
                                                                ?: 0.0, radarInfo.first.value.GetSouthWestCorner()?.GetLongitude()?.toDouble()
                                                                ?: 0.0)))
                                                mMapController.SetCameraPosition(radarInfo.first.value.GetLatitude()?.toDouble()
                                                        ?: 0.0, radarInfo.first.value.GetLongitude()?.toDouble()
                                                        ?: 0.0, 5.0)
                                            }
                                        }
                                    }
                                }
                            })

                            mRadarImageDataProvider.GetRadarAnimationObservable().observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<List<RadarDataController.ImageInfo>> {
                                override fun onComplete() {}
                                override fun onSubscribe(d: Disposable) {
                                    mImageListDisposable?.dispose()
                                    mImageListDisposable = d
                                }

                                override fun onNext(inImageList: List<RadarDataController.ImageInfo>) {
                                    mHoldPlay = false
                                    mMapController.SetFrameCount(inImageList.size)

                                    (mBinding?.playPauseButton as View?)?.visibility = View.VISIBLE
                                }

                                override fun onError(e: Throwable) {
                                    e.printStackTrace()
                                    SetSnackbarText("Error", "", "")
                                }
                            })



                            Observables.combineLatest(mMapController.GetActiveImageIndexObservable(),
                                    mRadarImageDataProvider.GetSiteInfoObservable(),
                                    mRadarImageDataProvider.GetRadarAnimationObservable())
                            { activeImage, radarInfo, imageList ->

                                Pair(imageList[activeImage], radarInfo)
                            }.observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<Pair<RadarDataController.ImageInfo, Map.Entry<String, RadarDetails>>> {
                                override fun onComplete() {}
                                override fun onSubscribe(d: Disposable) {
                                    mActiveImageDisposable?.dispose()
                                    mActiveImageDisposable = d
                                }

                                override fun onNext(pair: Pair<RadarDataController.ImageInfo, Map.Entry<String, RadarDetails>>) {
                                    if (!pair.first.GetImage().isRecycled) {
                                        if (!mHoldPlay)
                                            mRadarImageSource?.setImage(pair.first.GetImage())
                                        if(mState == PageStateInfo.PageState.kData)
                                            SetSnackbarText(pair.second.key, pair.second.value.GetName(), if (mHoldPlay) "" else pair.first.GetTimestring())
                                    }
                                }

                                override fun onError(e: Throwable) {
                                    SetSnackbarText("Error", "", "")
                                    e.printStackTrace()
                                }
                            })
                        }
                    }
                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                }

            })
        }

        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE)
            mBinding?.overlayOpacityText?.visibility = View.VISIBLE
    }


    override fun onPause() {
        mRadarImageDataProvider.OnPause()

        mPageStateDisposable?.dispose()

        mActiveImageDisposable?.dispose()
        mImageListDisposable?.dispose()

        mPageStateDisposable = null
        mActiveImageDisposable = null
        mImageListDisposable = null

        super.onPause()
    }

    override fun onDestroyView() {
        mBinding?.playPauseButton?.Dispose()
        super.onDestroyView()
    }


    override fun onDestroy() {
        mMapController.Dispose()

        mTransparencyDisposable?.dispose()
        mSelectedRadarDisposable?.dispose()
        mCameraUpdateDisposable?.dispose()

        mTransparencyDisposable = null
        mSelectedRadarDisposable = null
        mCameraUpdateDisposable = null

        mEmtpyImage.recycle()

        super.onDestroy()
    }


    override fun Close() {
        Observable.create(ObservableOnSubscribe<Void>{
            RevealView(mBinding?.radarLayout, mBinding?.childFragmentContainer)
            mBinding?.readingInfoLayout?.visibility = View.VISIBLE
            (mBinding?.playPauseButton as View?)?.visibility = View.VISIBLE
            it.onComplete()
        }).subscribeOn(AndroidSchedulers.mainThread()).subscribe(object: Observer<Void> {
            override fun onComplete() {}
            override fun onSubscribe(d: Disposable) {}
            override fun onNext(t: Void) {}
            override fun onError(e: Throwable) {
                e.printStackTrace()
            }
        })
    }


    private fun RevealView(inClickView: View?, inViewToReveal: View?) {
        val cx = inClickView?.left ?: 0 + (inClickView?.right ?: 0) - 24
        val cy = (inClickView?.top ?: 0 + (inClickView?.bottom ?: 0)) - 24
        val radius = Math.max(inViewToReveal?.width ?: 0, inViewToReveal?.height ?: 0) * 2.0f

        if (inViewToReveal?.visibility != View.VISIBLE) {
            inViewToReveal?.visibility = View.VISIBLE
            val hide = ViewAnimationUtils.createCircularReveal(inViewToReveal, cx, cy, 0f, radius)
            hide.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    inViewToReveal?.visibility = View.VISIBLE
                    hide.removeListener(this)
                }
            })
            hide.start()

        } else {
            val reveal = ViewAnimationUtils.createCircularReveal(
                    inViewToReveal, cx, cy, radius, 0f)
            reveal.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    inViewToReveal.visibility = View.INVISIBLE
                    reveal.removeListener(this)
                }
            })
            reveal.start()
        }
    }



    internal fun SetSnackbarText(inRadarId: String?, inRadarName: String?, inTimeString: String?)
    {
        if(isAdded && !isDetached) {
            var timeString = ""

            if (inTimeString != null) {
                if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT)
                    timeString = "<br>"
                else
                    timeString += "&nbsp;&nbsp;&nbsp; "
            }

            timeString += inTimeString ?: ""

            val radarId = inRadarId ?: ""
            val radarName = inRadarName ?: ""

            if(Build.VERSION.SDK_INT > Build.VERSION_CODES.M)
                mBinding?.readingInfoTextView?.text = Html.fromHtml("<b><span style=\"color: #acc0fb;\">$radarId</span>&nbsp;&nbsp; $radarName</b>$timeString\n", 0)
            else
                mBinding?.readingInfoTextView?.text = Html.fromHtml("<b><font color=\"#acc0fb\">$radarId</font>&nbsp;&nbsp; $radarName</b>$timeString\n")
        }
    }
}
